src/mds/Locker.cc

```diff
@@ -225,6 +225,37 @@ struct MarkEventOnDestruct {
   }
 };
 
+/*
+ * Try to detect the possible deadlock between requests, and
+ * will try to break the newer request.
+ */
+void Locker::detect_dead_locks(const MDRequestRef& mdr, int lock_type)
+{
+  bool any_waiter = false;
+  bool order_mess = false;
+
+  dout(15) << "detect_dead_locks " << *mdr << dendl;
+  for (auto it = mdr->locks.begin(); it != mdr->locks.end(); ) {
+    SimpleLock *lock = it->lock;
+
+    if (lock->is_waiter_for(SimpleLock::WAIT_ALL)) {
+      any_waiter = true;
+    }
+    if (lock->get_type() > lock_type) {
```

Comment:
You need to check the lock parent is the same as well.

Reply:
@batrick Will this make sense ?

```
#define CEPH_LOCK_DN          (1 << 0)
#define CEPH_LOCK_DVERSION    (1 << 1)
#define CEPH_LOCK_IQUIESCE    (1 << 4)  /* mds internal */
#define CEPH_LOCK_ISNAP       (1 << 5)  /* snapshot lock. MDS internal */
#define CEPH_LOCK_IPOLICY     (1 << 6)  /* policy lock on dirs. MDS internal */
#define CEPH_LOCK_IFILE       (1 << 7)
#define CEPH_LOCK_INEST       (1 << 8)  /* mds internal */
#define CEPH_LOCK_IDFT        (1 << 9)  /* dir frag tree */
#define CEPH_LOCK_IAUTH       (1 << 10)
#define CEPH_LOCK_ILINK       (1 << 11)
#define CEPH_LOCK_IXATTR      (1 << 12)
#define CEPH_LOCK_IFLOCK      (1 << 13)  /* advisory file locks */
#define CEPH_LOCK_IVERSION    (1 << 14)  /* mds internal */
```

The `_DN` lock's parent will be different with the other always, right ?

Such as for the deadlock issue in `[ceph-users] MDS Behind on Trimming...`, from current block ops output it's a deadlock between `CInode's` `_IFILE` or `_INEST` and a `_DN`.

Reply:
No, I'm referring to `SimpleLock::parent` which is the CInode/CDentry the lock is protecting. You need to ensure that the parent is the same when comparing the locks.

Reply:
Let's me explain more about this.

For example firstly a request acquires:

```
in->filelock
in->authlock
```

Then later in another `acquire_locks()` when trying to acquire the `dn->lock`, if we check the parents of the locks then the order will be:

```
in->filelock
in->authlock
dn->lock
```

But actually it should be 

```
#define CEPH_LOCK_DN (1 << 0) 
#define CEPH_LOCK_IFILE (1 << 7) 
#define CEPH_LOCK_IAUTH (1 << 10)
```

So IMO we should ignore the parents check anyway.

Did I miss anything here ? Hope we are on the same page :-)

Reply:
The locking order is not simply the lock type. It is documented here:

https://github.com/ceph/ceph/blob/367a84918bb3cacbc3a7fac5dbfed468b73cf523/src/mds/MDCache.cc#L8264-L8281

but I'm unsure if that captures all of the implicit (by code) rules.